std/typecons.d

```diff
@@ -1135,12 +1135,20 @@ if (distinctFieldNames!(Specs))
             Returns:
                 A `size_t` representing the hash of this `Tuple`.
          */
-        size_t toHash() const nothrow @trusted
+        size_t toHash()() const nothrow @trusted
```

Comment:
Looks like why can't change this to a template:

```
std/functional.d(1003): Error: AA key type Tuple should have 'size_t toHash() const nothrow @safe' if opEquals defined
generated/linux/debug/64/publictests/std_functional.d(193): Error: template instance std_functional.__unittest_generated_linux_debug_64_publictests_std_functional_d_187_15.memoize!(fib) error instantiating
std/functional.d(1003): Error: AA key type Tuple should have 'size_t toHash() const nothrow @safe' if opEquals defined
generated/linux/debug/64/publictests/std_functional.d(205): Error: template instance std_functional.__unittest_generated_linux_debug_64_publictests_std_functional_d_199_16.memoize!(fact) error instantiating
std/functional.d(1003): Error: AA key type Tuple should have 'size_t toHash() const nothrow @safe' if opEquals defined
generated/linux/debug/64/publictests/std_functional.d(219): Error: template instance std_functional.__unittest_generated_linux_debug_64_publictests_std_functional_d_211_17.memoize!(factImpl) error instantiating
```

How about simply removing the method for `betterC` for now?

Reply:
How about this:
```D
size_t toHash() const nothrow @trusted
{
    version(D_BetterC)
    {
        // Disabled in -betterC for now as TypeInfo isn't present
        assert(0, "Not implemented");
    }
    else
    {
        size_t h = 0;
        foreach (i, T; Types)
            h += typeid(T).getHash(cast(const void*)&field[i]);
        return h;
    }
}
```

Reply:
That would fail at runtime which is strictly worse then compile-time failure.

Reply:
Of course, that's why my first suggestion was to make `toHash` a template function and use `static assert`. Thinking about it more I don't see why we define our own `opEquals` and `toHash`, instead of relying on the compiler generated ones.